src/Spago/Config/AST.hs

```diff
@@ -0,0 +1,964 @@
+          Dhall.TextLit (Dhall.Chunks [] txt) -> pure $ Just txt
+          other -> do
+            logWarn $ display $ "not listlit: " <> failedToAddDepsExpectedRecordKey other
+            pure Nothing
+        _ -> do
+          logWarn $ display $ "no record lit: " <> failedToAddDepsExpectedRecordKey (Dhall.RecordLit kvs)
+          pure Nothing
+      other -> do
+        logWarn $ display $ "other: " <> failedToAddDepsExpectedRecordKey other
+        pure Nothing
+      where
+        failedToAddDepsExpectedRecordKey e =
+          "Failed to add dependencies. You should have a record with the `" <> key <> "` key for this to work.\n" <>
+          "Expression was: " <> pretty e
+
+-- | "dependencies" - Reduce chance of spelling mistakes/typos
+dependenciesText :: Text
+dependenciesText = "dependencies"
+
+-- | "sources" - Reduce chance of spelling mistakes/typos
+sourcesText :: Text
+sourcesText = "sources"
+
+-- | "name" - Reduce chance of spelling mistakes/typos
+nameText :: Text
+nameText = "name"
+
+-- | Indicates where we are in the expresion
+data ExprLevel
+  = AtRootExpression Text
+  -- ^ The outermost expression
+  --   that indicates which field within the record
+  --   expression we want to update.
+  | SearchingForField !(NonEmpty Text)
+  -- ^ An expression within the root expression
+  --   where we still haven't found the record with the
+  --   `dependencies` field key.
+  --   The list of Text values is a stack of keys. We use
+  --   the key on the head of the list when we encounter a
+  --   `RecordLit` to loookup a field within that record.
+  --   `Field` pushes new keys on top of the stack
+  --   `RecordLit` looks up those fields and drops the keys on the stack
+  --   Transitioning from `AtRootExpression` to `SearchingForField`
+  --   should therefore always have `"dependencies"` at the bottom of the stack.
+  | WithinField
+  -- ^ This expression is the value associated with the `dependencies` field.
+  --   It must be a `List`-like structure, so immediately append the new packages
+  --   to the expression via a `ListAppend` or merge `ListLit`s together
+  --   if it's a `ListAppend`.
+  deriving (Eq, Show)
+
+data UpdateResult
+  = Updated !Expr
+  -- ^ The expression was succesfully updated with the new packages
+  | VariableName !(Text, Int)
+  -- ^ The expression to update is the binding value with this name
+  --   that corresponds to the specified de Brujin index
+  | EncounteredEmbed
+  -- ^ We encountered an `Embed` constructor. We can make only
+  --   two assumptions about it:
+  --    1. If we are `WithinField`, then this must be an
+  --       expression that produces a `List`-like structure.
+  --    2. We are NOT `WithinField`, then this must be an
+  --       expression that produces a `Record`-like Config-schema structure.
+  --       Thus, a "dependencies" key should exist. If it doesn't, the
+  --       Dhall expression is invalid (as verified by our prior normalized expression)
```

Comment:
I don't understand why are we able to make these assumptions?

Reply:
I've updated this doc comment, but here's my understanding. If I'm misunderstanding how Dhall's `normalize` function works, then perhaps I'm making a mistake here.

If we take a Dhall expression (e.g. what's found in `./spago.dhall`) and normalize it, it gets rid of all the non-literal constructors (e.g. `ListAppend`, `Prefer`, etc.) and replaces them all with literal values (e.g. `ListLit`, `RecordLit`, etc.). For example, an expression like `["foo"] # ["bar"]` becomes `["foo", "bar"]`. Similarly (and this is my understanding), an expression like `./spago.dhall with name = "foo"` where `./spago.dhall`'s file content is `{ name = "bar", ... }` will produce `{name = "foo", ... }` where `...` are the same values stored in the `./spago.dhall` file.

So, if we have an expression like this:
```dhall
{ name = "foo", dependencies = ./spago-deps.dhall }
-- which normalizes to
{ name = "foo", dependencies = ["foo"] }
```
then [here](https://github.com/purescript/spago/pull/815/files#diff-560aba1d76fb29566dd1ccf7e6fabe4e1cc9d5c011606bfd6de21485b70fe8cfR88), where we are looking up the text values for the "dependencies" key in the normalized version of the expression, [will succeed and return `Just`](https://github.com/purescript/spago/pull/815/files#diff-560aba1d76fb29566dd1ccf7e6fabe4e1cc9d5c011606bfd6de21485b70fe8cfR128) if `./spago-deps.dhall` produces a `List Text` Dhall expression and will fail otherwise.

Knowing that in advance, if I was updating the `dependencies` key, was at the `WithinField` level, and encountered an `Embed` case, then I know the `Embed` case must produce a `List Text` expression. Thus, I can wrap that expression in a `ListAppend` even though I don't know what madness may lie behind that `Embed`.

Similarly, if I have an expression like this:
```dhall
{ outer = { config = ./spago.dhall } }.outer.config
-- which normalizes to
{ name = "foo", dependencies = [], sources = [], ... }
```
then, because the normalized expression produces a `RecordLit` where the `dependencies` key stores a `List Text` (i.e. a valid `Config` dhall expression, at least for updating the `dependencies` field), here's how my `modifyRawAST'`/`modifyRawDhallExpression` would work:

1. Make a request to update the `dependencies` field by calling `modifyRawDhallExpression` with `InsertListText "new" with level `AtRootExpression "dependencies"`.
1. Found a `Field` that selects the `config` field. Before we can find a record expression and lookup its `dependencies` field, we must first find a record expression that has a `config` field. Thus, we now transition to `SearchingForField ("config" :| ["dependencies"])`)
1. Found a `Field` that selects the `outer` field. Same as before. Thus, we now transition to `SearchingForField ("outer" :| ["config", "dependencies"])`)
1. Found a `RecordLit` and looking up the `outer` field was successful. Transition to `SearchingForField ("config" :| ["dependencies"])`
1. Found a `RecordLit` and lookingup the `config` field was successful. Transition to `SearchingForField ("dependencies" :| [])`
1. Found an `Embed`. Since we're not at `WithinField` yet, then this MUST be a record expression that stores a `Config`-like expression somewhere in it that we can find by using the rest of the keys on our stack. Otherwise, the normalization step on the original expression would have returned `Nothing` rather than `Just`. Since the keys in our stack are just `"dependencies :| []`, then this `Embed` IS a `Config`-like Dhall expression.
1. Since this is a record expression, and we're updating the `dependencies` field, we need to build off of the record expression produced by the `Embed`. Thus, we will wrap it in a `Let` binding to have access to it, and override its `dependencies` value with the record expression's original `dependencies` value plus the new changes.
1. We now return `Just Updated` and rewrap all the other expressions around this one. 

Thus...
```dhall
-- start
{ outer = 
    { config = ./spago.dhall 
    } 
}.outer.config

-- final expression produced
{ outer = 
    { config = 
        let __embed = ./spago.dhall 
        in  __embed 
                with dependencies = __embed.dependencies # ["new"] 
    }
}.outer.config
```